### ISSUE TITLE
Very Slow Inference on CPU 

### ISSUE THREAD

**Issue by AIrocker** [Daniil-Osokin/lightweight-human-pose-estimation.pytorch#32]
Hi  I found the model is very very slow on CPU using the pretrained weight "checkpoint_iter_370000.pth". I have attached the code below.  I have tested different scenarios, and summarize the results below:
GPU w pretrained weight:   0.007 sec 
GPU w/o pretrained weight: 0.007 sec 
**CPU w pretrained weight: 2.829 sec** 
CPU w/o pretrained weight: 0.376 sec 

Could you kindly explain why the inference time using CPU and pretrained weight is so slow ? 

```
import argparse
import cv2
import numpy as np
import torch
from models.with_mobilenet import PoseEstimationWithMobileNet
from modules.keypoints import extract_keypoints, group_keypoints
from modules.load_state import load_state
from modules.pose import Pose, propagate_ids
import time
#from val import normalize, pad_width
device = torch.device('cpu')
model_Mobilenet = PoseEstimationWithMobileNet().to(device)
checkpoint = torch.load('checkpoint_iter_370000.pth', map_location=lambda storage, loc: storage)
load_state(model_Mobilenet, checkpoint)
model_Mobilenet.eval()
input = torch.Tensor(2, 3, 368, 368).to(device)

since = time.time()
stages_output= model_Mobilenet(input)
PAF_Mobilenet, Heatmap_Mobilenet = stages_output[-1], stages_output[-2]
print('Mobilenet Inference time is {:2.3f} seconds'.format(time.time() - since))
```

**Comment by Daniil-Osokin** [comment]
If you have used for time measurements code from above, then the reason might be in small statistic. You know, that cpu executes not only network inference, but also OS and other background tasks at the same time, so there can be some delays. Better to run number experiments in a loop (e.g. 100 iterations), measure inference time at each iteration, then average it or take median. This network has determenistic inference, so the complexity (flops number) should be the same regardeless of weights choice. And for the fastest inference, please consider [c++ demo](https://github.com/Daniil-Osokin/lightweight-human-pose-estimation.pytorch#cpp-demo).

**Comment by AIrocker** [comment]
Hi Daniil,  thanks for your kindly response.  Follow your instruction, I run the code in a 100 iteration loop to obtain an average inference time, below is the summary table. The results are reproduced using other desktops. 
As you explained, the flops number should be the same regardless of weights choice.  Really confused why the inference time soars with CPU and pretrained weight.

  

|  Engine     |   Weight   |   Inference Time |  
|:-------------:|:------------:|:--------------------:|
|CPU           |  pretrain   |  3.53                  |  
|CPU           |  random initial       |  0.32                  |  
|GPU           |  pretrain   |  0.057                |  
|GPU           |  random initial       |  0.057                |   

```
import cv2
import numpy as np
import torch

from models.with_mobilenet import PoseEstimationWithMobileNet
from modules.keypoints import extract_keypoints, group_keypoints
from modules.load_state import load_state
from modules.pose import Pose, propagate_ids
import time
#from val import normalize, pad_width
#device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
device = torch.device('cpu')

model_Mobilenet = PoseEstimationWithMobileNet().to(device)
checkpoint = torch.load('checkpoint_iter_370000.pth', map_location=lambda storage, loc: storage)
load_state(model_Mobilenet, checkpoint)
model_Mobilenet.eval()

input = torch.Tensor(2, 3, 368, 368).to(device)

since = time.time()

for i in range(101):
    stages_output= model_Mobilenet(input)

print('Mobilenet Inference time is {:2.3f} seconds'.format((time.time() - since)/100))
```
I also have a close-look at the pretrained weights. Below shows the parameters for first conv layer. The parameters are very small i.e. 6e-42.   Not sure if that is the reason causing high inference time using CPU ? 

```
params = list(model_Mobilenet.parameters())
print(params[0])
tensor([[[[ 1.3548e-41, -2.8558e-42,  7.1985e-42],
          [ 6.8509e-42, -1.2022e-41, -1.3971e-42],
          [-9.6213e-42,  7.1340e-42,  1.3549e-41]],

         [[-5.7888e-42,  1.0885e-41,  1.9870e-42],
          [-6.3605e-42, -5.6192e-42,  5.8098e-42],
          [-1.0887e-41,  8.5269e-42,  7.8038e-42]],

         [[-4.8905e-42, -5.1245e-42, -9.7853e-42],
          [ 8.7847e-42, -2.4144e-42,  2.8054e-42],
          [-8.5423e-42, -3.8802e-42,  6.0704e-42]]],


        [[[-4.6523e-43, -1.6549e-42, -5.2633e-42],
          [ 1.2619e-41,  1.2462e-41, -1.0759e-41],
          [ 7.3022e-42,  9.8735e-42, -1.0707e-41]],

         [[-1.0741e-41,  5.7719e-42, -1.0706e-41],
          [ 1.1174e-41, -4.2726e-42,  2.0473e-42],
          [-1.0983e-41,  9.1477e-42, -9.6227e-42]],
```

**Comment by Daniil-Osokin** [comment]
Does this reproduced with real image, not generated tensor?

**Comment by AIrocker** [comment]
Daniil,  Yes.  It is reproduced with real image.  I rounded over the pretrained weight parameters by 4 decimals. The inference speed is back to 0.3 sec.  I didn't find any accuracy drop.  For your information.  Not sure if you can reproduce this scenario though.  Wonder if there is any theoretical explanation on this ?  

```
params = list(model_Mobilenet.parameters())
print('the length of parameters is', len(params))
for i in range(len(params)):
    params[i].data = torch.round(params[i].data*10**4) / 10**4
```

**Comment by Daniil-Osokin** [comment]
There can be some exceptions which slow down the work, when cpu operates with small numbers (around machine zero). E.g. working with [denormals](https://en.wikipedia.org/wiki/Denormal_number) is usually slower, or some optimized code may expect numbers from specific range and, if numbers fall out this range, the non-optimized code runs.

**Comment by Daniil-Osokin** [comment]
This can be closed.
